### ISSUE TITLE
Various reset parameter

### ISSUE THREAD

**Issue by dora-gt** [Unity-Technologies/ml-agents#107]
Hi,

When I implement curriculum, sometimes I want to use int reset parameter while current ml-agents only accepts float values.

https://github.com/Unity-Technologies/ml-agents/blob/master/unity-environment/Assets/ML-Agents/Scripts/Academy.cs#L60

People may want to use various types of reset parameters.

**Comment by vincentpierre** [comment]
Hi, it is true we only allowed passing floats as reset parameters. If you want to pass an integer, you can cast the float into an int in C#. The dictionary `resetParameters` is from string to float, it would not be as easy to use if it was a dictionary from string to anything.

**Comment by dora-gt** [comment]
Thank you for the reply.

Yes, I knew I could cast it, but I was upset of casting because it is not accurate.
I was thinking something like POCO. It is easy to use (though it may cost more).

Thank you anyway.

**Comment by vincentpierre** [comment]
What do you mean by "casting is not accurate" ? If you send the reset parameter with a value 1.0f casting it to int will give you 1. You will need to make sure the reset parameter you send in python has value 1. What operation are you doing ?

**Comment by dora-gt** [comment]
For example, if the environment has various flags,

- The environment has XX
- The environment has YY
- The environment has ZZ

and I want to pass it as reset parameter, I would use flags like...

`flag = XX | YY | ZZ`

and if the environment is really complex, I may use fully 32bit.
Then what happens?

I don't know much about Python, but there may be int, and it may accurately express the flag.
Then passing it to C# through JSON (maybe? I haven't read the code) and cast the string to float, and casting it again to int, the flag may lost, having different meaning.

I think it is not accurate.

**Comment by dora-gt** [comment]
P.S.

Because I thought so, I'm using reset parameter only as "curriculum_id". It doesn't matter because the value is really small. Although... the architecture may not be beautiful because the setting of environment depends on C# code, and setting is dispersed.

We need beautiful architecture to implement curriculum training but I don't know the answer 😅

**Comment by vincentpierre** [comment]
There is something I do not understand. Are you using 3 reset parameters? If you have resetParameter XX = 1.0f then X is true else false. This way, you can convert float to bool directly. I suppose XX | YY | ZZ means concatenation but why not use use 3 different flags as reset parameters ? I think I am missing something...

**Comment by dora-gt** [comment]
OK, if I use 3 independent reset parameters, there isn't any problem because, as you say, I can pass 3 float variables of 1.0f or 0.0f meaning boolean and I can cast it accurately.

On the other hand, passing many variables look a bit uncool because

- data increases (JSON may have many strings as the names)
  - like...
    - { param1: 1.0f, param2: 1.0f, param3: 0.0f, ........., param32: 1.0f }
  - using int flag like...
    - { flags: 2147483647 }
- setting many fields on inspector is tiresome

So, I felt that if I could set int as reset parameters.

**Comment by SPINACHCEO** [comment]
Hi @dora-gt 
I think you can very easily cast float into int with
`(int)academy.resetParameters["your param"]`
If from python the float is 1.0f after the cast it will become 1.

It is totally accurate. :)

**Comment by dora-gt** [comment]
Hi @SPINACHCEO ,

Yes, I know it works but sometimes, for example using it as flag, it can be not accurate like I wrote above. To avoid it, we have to define many reset parameters and it means data increases. I think it is tiresome... That is the point.

Here is what I mean "not accurate"
```
int flag = 0;
for (int x = 0; x < 31; x ++)
{
    flag = flag | (1 << x);
}
// here, flag means fully filled 31 bit like (01111111111111111111111111111111)

string flagInJson = flag.ToString();
float flagInCS = float.Parse(flagInJson);

// this is value from Python through JSON
int castedFlag = (int)flagInCS;

Debug.Log(string.Format("flag: {0}, flagInJson:{1}, flagInCS: {2}, castedFlag: {3}", flag, flagInJson, flagInCS, castedFlag));
```
↓
`flag: 2147483647, flagInJson:2147483647, flagInCS: 2.147484E+09, castedFlag: -2147483648`

**Comment by vincentpierre** [comment]
Looks like this happens because the float is too big so some information is lost starting in `flagInCS`. I will see what can be done in the code, in the meantime, I would send 2 floats as flags to avoid sending a float too big.

**Comment by dora-gt** [comment]
Thank you but please put priority on your more important work, I can work around.

**Comment by mmattar** [comment]
Closing this out due to inactivity. We've also since released v0.3 which includes several new enhancements. @dora-gt If you run into any issues or have any proposals for features, please create a new Issue with the 'request' label.

**Comment by dora-gt** [comment]
Thank you for the follow-up. I'll check v0.3.
